### PR TITLE
Refactor/input handling

### PR DESCRIPTION
**Pull request by Consolatis**
Nothing urgent but we accumulated quite a bit of input handling all over the place.

So this PR tries to slightly refactor the input handling parts of labwc into `src/input/`.
It also splits off the gesture handling from `src/cursor.c` into `src/input/gestures.c` and makes the keyboard key / modifier listeners private.

The implicit UAF condition when finalizing the touch (and now gesture) handlers before the cursor one is dealt with by just having `cursor_finalize()` finalize those itself (and `cursor_init()` init them).

I've tested a rebase of #626 on top of this and it went through without issues.

This is a Draft for now (and `[wip]` commits) to ask for opinions about the approach first.
It will come in handy when restarting work on #1060 as that adds yet another input method.

@johanmalm @jlindgren90

### CODE REVIEW

File: include/input/keyboard.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+struct seat;
+struct keyboard;
+struct wlr_keyboard;
+
+void keyboard_init(struct seat *seat);
+void keyboard_finish(struct seat *seat);
+
+void keyboard_set_numlock(struct wlr_keyboard *keyboard);
+void keyboard_add_keyboard_handler(struct keyboard *keyboard, struct wlr_keyboard *wlr_kb);
```
Comment (Consolatis) [inline include/input/keyboard.h:15]: It should likely be enough to just use the first arg to get the `wlr_keyboard`. Verify.

File: include/labwc.h [hunk 2]
```diff
@@ -30,7 +30,6 @@
 #include <wlr/types/wlr_relative_pointer_v1.h>
 #include <wlr/types/wlr_pointer.h>
 #include <wlr/types/wlr_pointer_constraints_v1.h>
-#include <wlr/types/wlr_pointer_gestures_v1.h>
```
Comment (Consolatis) [inline include/labwc.h:33]: All of these could likely be moved to `include/input/cursor.h`. Verify.

**Comment by jlindgren90** [comment]
Very nice, LGTM but didn't test.

An idea for some follow-up refactoring: could we split up `struct seat` into some smaller structs, maybe following similar divisions to what you have here (cursor/gestures/keyboard/etc.)? Doesn't need to be done immediately or in this PR, just something to consider.

**Comment by Consolatis** [comment]
> An idea for some follow-up refactoring: could we split up `struct seat` into some smaller structs

I think we could start with moving `struct seat` from `include/labwc.h` into `include/seat.h` (same for the public `seat_` functions).
Same for things like `enum input_mode` and `struct input`.

We could also consider a `include/output.h`. Basically trying to get rid of that giant `labwc.h` include bit by bit.

**Comment by Consolatis** [comment]
Rebased on master + renamed new keyboard function into  `keyboard_setup_listeners()` + cleaned up the commits .
#626 can still be rebased fine on top of this branch.

As we are potentially doing some changes to the input handling soon, getting this PR merged first seems easier.
I am not sure how exactly `git mv` reacts to being rebased on top of something that changed the contents of the file being moved.

**Comment by jlindgren90** [comment]
I haven't looked over the latest changes but I agree with getting this merged. I don't think we should delay changes to master for the sake of the wlroots 0.17 PR/branch.

**Comment by johanmalm** [comment]
Thanks. Nice!

**Pull request merged**